### PR TITLE
Add invariant that an object with modified properties must be valid.

### PR DESCRIPTION
**Pull request by NTillmann**
Release notes: None

The uncovered bug is tracked in #2279.

### CODE REVIEW

File: src/values/ObjectValue.js [hunk 1]
```diff
@@ -325,6 +325,11 @@ export default class ObjectValue extends ConcreteValue {
   // never be serialized.
   refuseSerialization: boolean;
 
+  // Checks whether effects are properly applied.
```
Comment (cblappert) [inline src/values/ObjectValue.js]: I don't see anything that ensures that _isPartial is only true when effects are applied/undone. I thought this was an annotation used in our models/used for snapshotting.

**Comment by hermanventer** [comment]
I'm looking into the bug.

**Comment by hermanventer** [comment]
I've verified that this invariant is valid with #2402.

**Comment by trueadm** [comment]
I updated the PR with master and added the check in a bunch of other places. It seems we are hitting more cases now, which is good in some respects. Herman's join PR seems to fix all the failing test cases though.

**Pull request closed without merging**